tls/client-tls-cryptocb.c

```diff
@@ -31,28 +31,115 @@
 #include <unistd.h>
 
 /* wolfSSL */
-#include <wolfssl/options.h>
+#ifndef WOLFSSL_USER_SETTINGS
+    #include <wolfssl/options.h>
+#endif
 #include <wolfssl/ssl.h>
 #include <wolfssl/wolfcrypt/sha256.h>
 #include <wolfssl/wolfcrypt/cryptocb.h>
 #include <wolfssl/wolfcrypt/error-crypt.h>
 
 #define DEFAULT_PORT 11111
 
-#define CA_FILE "../certs/ca-cert.pem"
+#define USE_ECDHE_ECDSA
+#define USE_TLSV13
+
+#ifdef USE_ECDHE_ECDSA
+#define CERT_FILE   "../certs/client-ecc-cert.pem"
+#define KEY_FILE    "../certs/ecc-client-key.pem"
+#define CA_FILE     "../certs/ca-ecc-cert.pem"
+#else
+#define CERT_FILE   "../certs/client-cert.pem"
+#define KEY_FILE    "../certs/client-key.pem"
+#define CA_FILE     "../certs/ca-cert.pem"
+#endif
 
 #ifdef WOLF_CRYPTO_CB
 /* Example custom context for crypto callback */
 typedef struct {
     int exampleVar; /* example, not used */
 } myCryptoCbCtx;
 
-static void error_out(char* msg, int err)
+typedef struct {
+    word32 bufSz;
+} hash_ctx_t;
+
+#ifdef USE_OPENSSL
+#include <openssl/sha.h>
+#endif
+
+/* type: WC_HASH_TYPE_SHA, WC_HASH_TYPE_SHA256, WC_HASH_TYPE_SHA384, etc */
+/* in: Update (when not NULL) / Final (when NULL) */
+static int cb_hash(int type, const byte* in, word32 inSz, byte* digest,
+    void* shactx, void** devCtx, word32 flags)
 {
-    printf("Failed at %s with code %d\n", msg, err);
-    exit(1);
-}
+    int ret = 0;
+    enum wc_HashType hash_type = (enum wc_HashType)type;
+    hash_ctx_t* ctx = (hash_ctx_t*)*devCtx;
+    byte*  hashBuf = NULL;
+    word32 hashBufSz = 0;
+
+    /* for updates alloc/realloc and copy */
+    if (in != NULL) {
+        if (ctx == NULL) {
+            ctx = (hash_ctx_t*)malloc(sizeof(hash_ctx_t) + hashBufSz + inSz);
```

Comment:
Have two implementations, USE_OPENSSL that allocates and other that has the wc_HashAlg as the devCtx?
That way it shows the two different ways of doing it.